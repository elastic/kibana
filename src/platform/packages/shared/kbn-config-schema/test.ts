import { TypeOf, schema } from '@kbn/config-schema';
import _ from 'lodash';

const testSchemaObject = schema.object({
    a: schema.string(),
    b: schema.number({ defaultValue: 5 }),
    c: schema.maybe(schema.boolean()),
});

const testSchemaAllOf = schema.allOf([
    testSchemaObject,
    schema.object({
        d: schema.string(),
        e: schema.number({ defaultValue: 3 }),
        f: schema.maybe(schema.boolean()) 
    }),
])

const testSchemaObjectInput = testSchemaObject.getInputSchema();
const testSchemaAllOfInput = testSchemaAllOf.getInputSchema();

type TestSchemaObjectInputType = typeof testSchemaObjectInput.type;
type TestSchemaAllOfInputType = typeof testSchemaAllOfInput.type;

const testObject: TestSchemaObjectInputType = {
    a: 'hello',
    // b is optional because it has a default value
    // c is optional because it's maybe
};

// complains that b is missing
