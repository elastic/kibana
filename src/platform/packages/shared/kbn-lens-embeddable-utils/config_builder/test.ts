import type { MetricState } from "./schema";

const test: MetricState = {
  type: 'metric',
  dataset: {
    type: 'index',
    index: 'test',
    time_field: 'test'
  },
  metric: {
    operation: 'count',

  },
  secondary_metric: {
    operation: 'min',
    field: 'test',
  },
  breakdown_by: {
    operation: 'terms',
    fields: ['test'],
    size: 5,
    increase_accuracy: true,
    collapse_by: 'avg',
  },
  ignore_global_filters: true,
  samplings: 1,
}