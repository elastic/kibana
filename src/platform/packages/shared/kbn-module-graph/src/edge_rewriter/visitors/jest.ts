/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the "Elastic License
 * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
 * Public License v 1"; you may not use this file except in compliance with, at
 * your election, the "Elastic License 2.0", the "GNU Affero General Public
 * License v3.0 only", or the "Server Side Public License, v 1".
 */

import { NodePath, VisitNodeFunction } from '@babel/traverse';
import { codeFrameColumns } from '@babel/code-frame';
import { VisitorContext, PluginState } from './types';
import { JestMockExpression } from '../../traverse/types';

/**
 * Checks if the preceding line contains a comment that opts out of jest mock rewriting
 */
function hasOptOutComment(path: NodePath<JestMockExpression>): boolean {
  const comments = path.parent.leadingComments;
  if (!comments || comments.length === 0) {
    return false;
  }

  // Check if any preceding comment contains the opt-out pattern
  return comments.some((comment) =>
    comment.value.trim().includes('kbn-dependency-resolver:keep-jest-mock')
  );
}

export function createJestVisitor<S extends PluginState>(
  visitorContext: VisitorContext
): VisitNodeFunction<S, JestMockExpression> {
  return function (path) {
    // Skip if engineer has opted out via comment
    if (hasOptOutComment(path)) {
      return;
    }

    const modulePath = path.node.arguments[0].value;
    const state = path.state;

    // Check if there's a rewrite available for this module
    visitorContext.withEdgeRewrite(null, modulePath, (rewrite) => {
      const code = state.file?.code || '';

      // Create helpful error message with suggestion
      const suggestion =
        `Consider updating the Jest mock to use the rewritten path:\n` +
        `  jest.mock('${rewrite.filePath}');\n\n` +
        `Or if you need to keep the original mock, add this comment on the preceding line:\n` +
        `  // kbn-dependency-resolver:keep-jest-mock`;

      const error = new Error(
        codeFrameColumns(code, path.node.loc!, {
          message: `Jest mock for '${modulePath}' should be updated to '${rewrite.filePath}'. ${suggestion}`,
        })
      );

      throw error;
    });
  };
}
