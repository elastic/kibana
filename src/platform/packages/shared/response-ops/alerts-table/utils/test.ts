/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the "Elastic License
 * 2.0", the "GNU Affero General Public License v3.0 only", and the "Server Side
 * Public License v 1"; you may not use this file except in compliance with, at
 * your election, the "Elastic License 2.0", the "GNU Affero General Public
 * License v3.0 only", or the "Server Side Public License, v 1".
 */

import { DeepPartial } from 'utility-types';
import MaybeMockedDeep = jest.MaybeMockedDeep;

/**
 * Utility function to create a partial object mock without losing type information
 *
 * @example
 * ```ts
 * const mockedObject = createPartialObjectMock<MyType>({ prop1: 'value' });
 * ```
 */
export const createPartialObjectMock = <T>(partialObject: DeepPartial<T>) =>
  partialObject as MaybeMockedDeep<T>;

/* eslint-disable no-console */
export const testQueryClientConfig = {
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
  logger: {
    log: console.log,
    warn: console.warn,
    error: () => {},
  },
};

export const getJsDomPerformanceFix = () => {
  const originalGetComputedStyle = Object.assign({}, window.getComputedStyle);

  return {
    fix: () => {
      // The JSDOM implementation is too slow
      // Especially for dropdowns that try to position themselves
      // perf issue - https://github.com/jsdom/jsdom/issues/3234
      Object.defineProperty(window, 'getComputedStyle', {
        value: (el: HTMLElement) => {
          /**
           * This is based on the jsdom implementation of getComputedStyle
           * https://github.com/jsdom/jsdom/blob/9dae17bf0ad09042cfccd82e6a9d06d3a615d9f4/lib/jsdom/browser/Window.js#L779-L820
           *
           * It is missing global style parsing and will only return styles applied directly to an element.
           * Will not return styles that are global or from emotion
           */
          const declaration = new CSSStyleDeclaration();
          const { style } = el;

          Array.prototype.forEach.call(style, (property: string) => {
            declaration.setProperty(
              property,
              style.getPropertyValue(property),
              style.getPropertyPriority(property)
            );
          });

          return declaration;
        },
        configurable: true,
        writable: true,
      });
    },
    cleanup: () => {
      Object.defineProperty(window, 'getComputedStyle', originalGetComputedStyle);
    },
  };
};
