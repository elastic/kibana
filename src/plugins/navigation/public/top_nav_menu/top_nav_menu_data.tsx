/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * and the Server Side Public License, v 1; you may not use this file except in
 * compliance with, at your election, the Elastic License or the Server Side
 * Public License, v 1.
 */

import { EuiButtonProps } from '@elastic/eui';

export type TopNavMenuAction = (anchorElement: HTMLElement) => void;

export interface TopNavMenuData {
  id?: string;
  label: string;
  run: TopNavMenuAction;
  description?: string;
  testId?: string;
  className?: string;
  disableButton?: boolean | (() => boolean);
  tooltip?: string | (() => string | undefined);
  emphasize?: boolean;
  iconType?: string;
  iconSide?: EuiButtonProps['iconSide'];
}

export interface RegisteredTopNavMenuData extends TopNavMenuData {
  appName?: string;
}
