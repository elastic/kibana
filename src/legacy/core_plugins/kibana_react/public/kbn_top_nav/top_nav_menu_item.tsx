/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import React, { Component } from 'react';
import { EuiButtonEmpty } from '@elastic/eui';

import { TopNavMenuData, CloseHandler, TopNavMenuAction } from './top_nav_menu_data';

interface Props {
  data: TopNavMenuData;
  onClick: (key: string, action: TopNavMenuAction, target?: any) => void;
}

interface State {
  isDisabled: boolean;
  closeHandler: CloseHandler;
}

export class TopNavMenuItem extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
  }

  private isDisabled(): boolean {
    const menuData = this.props.data;
    const val =
      typeof menuData.disableButton === 'function'
        ? menuData.disableButton()
        : menuData.disableButton;
    return val || false;
  }

  private handleClick() {
    const menuData = this.props.data;
    this.props.onClick(menuData.key, menuData.run, arguments[0].currentTarget);
  }

  public render() {
    const menuData = this.props.data;
    return (
      <EuiButtonEmpty
        size="xs"
        isDisabled={this.isDisabled()}
        onClick={this.handleClick.bind(this)}
      >
        {_.capitalize(menuData.label || menuData.key)}
      </EuiButtonEmpty>
    );
  }
}
