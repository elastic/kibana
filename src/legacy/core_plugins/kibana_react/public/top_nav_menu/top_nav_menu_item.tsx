/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { capitalize, isFunction } from 'lodash';
import React, { MouseEvent } from 'react';
import { EuiButtonEmpty, EuiToolTip } from '@elastic/eui';

import { TopNavMenuData } from './top_nav_menu_data';

export function TopNavMenuItem(props: TopNavMenuData) {
  function isDisabled(): boolean {
    const val = isFunction(props.disableButton) ? props.disableButton() : props.disableButton;
    return val!;
  }

  function getTooltip(): string {
    const val = isFunction(props.tooltip) ? props.tooltip() : props.tooltip;
    return val!;
  }

  function handleClick(e: MouseEvent<HTMLButtonElement>) {
    if (isDisabled()) return;
    props.run(e.currentTarget);
  }

  const btn = (
    <EuiButtonEmpty
      size="xs"
      isDisabled={isDisabled()}
      onClick={handleClick}
      data-test-subj={props.testId}
    >
      {capitalize(props.label || props.id!)}
    </EuiButtonEmpty>
  );

  const tooltip = getTooltip();
  if (tooltip) {
    return <EuiToolTip content={tooltip}>{btn}</EuiToolTip>;
  } else {
    return btn;
  }
}

TopNavMenuItem.defaultProps = {
  disableButton: false,
  tooltip: '',
};
