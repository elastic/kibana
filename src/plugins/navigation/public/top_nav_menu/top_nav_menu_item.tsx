/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import { upperFirst, isFunction } from 'lodash';
import React, { MouseEvent } from 'react';
import { EuiButtonEmpty, EuiToolTip } from '@elastic/eui';

import { EuiButton } from '@elastic/eui';
import { TopNavMenuData } from './top_nav_menu_data';

export function TopNavMenuItem(props: TopNavMenuData) {
  function isDisabled(): boolean {
    const val = isFunction(props.disableButton) ? props.disableButton() : props.disableButton;
    return val!;
  }

  function getTooltip(): string {
    const val = isFunction(props.tooltip) ? props.tooltip() : props.tooltip;
    return val!;
  }

  function handleClick(e: MouseEvent<HTMLButtonElement>) {
    if (isDisabled()) return;
    props.run(e.currentTarget);
  }

  const commonButtonProps = {
    isDisabled: isDisabled(),
    onClick: handleClick,
    iconType: props.iconType,
    iconSide: props.iconSide,
    'data-test-subj': props.testId,
    className: props.className,
  };

  const btn = props.emphasize ? (
    <EuiButton {...commonButtonProps} size="s" fill>
      {upperFirst(props.label || props.id!)}
    </EuiButton>
  ) : (
    <EuiButtonEmpty {...commonButtonProps} size="xs">
      {upperFirst(props.label || props.id!)}
    </EuiButtonEmpty>
  );

  const tooltip = getTooltip();
  if (tooltip) {
    return <EuiToolTip content={tooltip}>{btn}</EuiToolTip>;
  }
  return btn;
}

TopNavMenuItem.defaultProps = {
  disableButton: false,
  tooltip: '',
};
