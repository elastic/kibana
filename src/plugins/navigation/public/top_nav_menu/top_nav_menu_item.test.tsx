/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * and the Server Side Public License, v 1; you may not use this file except in
 * compliance with, at your election, the Elastic License or the Server Side
 * Public License, v 1.
 */

import React from 'react';
import { TopNavMenuItem } from './top_nav_menu_item';
import { TopNavMenuData } from './top_nav_menu_data';
import { shallowWithIntl } from '@kbn/test/jest';

describe('TopNavMenu', () => {
  const ensureMenuItemDisabled = (data: TopNavMenuData) => {
    const component = shallowWithIntl(<TopNavMenuItem {...data} />);
    expect(component.prop('isDisabled')).toEqual(true);

    const event = { currentTarget: { value: 'a' } };
    component.simulate('click', event);
    expect(data.run).toHaveBeenCalledTimes(0);
  };

  it('Should render and click an item', () => {
    const data: TopNavMenuData = {
      id: 'test',
      label: 'test',
      run: jest.fn(),
    };

    const component = shallowWithIntl(<TopNavMenuItem {...data} />);
    expect(component.prop('isDisabled')).toEqual(false);

    const event = { currentTarget: { value: 'a' } };
    component.simulate('click', event);
    expect(data.run).toBeCalledTimes(1);
    expect(data.run).toHaveBeenCalledWith(event.currentTarget);

    component.simulate('click', event);
    expect(data.run).toBeCalledTimes(2);
  });

  it('Should render item with all attributes', () => {
    const data: TopNavMenuData = {
      id: 'test',
      label: 'test',
      description: 'description',
      testId: 'test-class-name',
      disableButton: false,
      run: jest.fn(),
    };

    const component = shallowWithIntl(<TopNavMenuItem {...data} />);
    expect(component.prop('isDisabled')).toEqual(false);

    const event = { currentTarget: { value: 'a' } };
    component.simulate('click', event);
    expect(data.run).toHaveBeenCalled();
  });

  it('Should render emphasized item which should be clickable', () => {
    const data: TopNavMenuData = {
      id: 'test',
      label: 'test',
      iconType: 'beaker',
      iconSide: 'right',
      emphasize: true,
      run: jest.fn(),
    };

    const component = shallowWithIntl(<TopNavMenuItem {...data} />);
    const event = { currentTarget: { value: 'a' } };
    component.simulate('click', event);
    expect(data.run).toHaveBeenCalledTimes(1);
    expect(component).toMatchSnapshot();
  });

  it('Should render disabled item and it shouldnt be clickable', () => {
    ensureMenuItemDisabled({
      id: 'test',
      label: 'test',
      disableButton: true,
      run: jest.fn(),
    });
  });

  it('Should render item with disable function and it shouldnt be clickable', () => {
    ensureMenuItemDisabled({
      id: 'test',
      label: 'test',
      disableButton: () => true,
      run: jest.fn(),
    });
  });

  it('Should render disabled emphasized item which shouldnt be clickable', () => {
    ensureMenuItemDisabled({
      id: 'test',
      label: 'test',
      iconType: 'beaker',
      iconSide: 'right',
      emphasize: true,
      disableButton: true,
      run: jest.fn(),
    });
  });

  it('Should render emphasized item with disable function and it shouldnt be clickable', () => {
    ensureMenuItemDisabled({
      id: 'test',
      label: 'test',
      iconType: 'beaker',
      iconSide: 'right',
      emphasize: true,
      disableButton: () => true,
      run: jest.fn(),
    });
  });
});
