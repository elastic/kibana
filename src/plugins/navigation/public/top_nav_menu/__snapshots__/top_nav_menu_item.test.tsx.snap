// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopNavMenu Should render emphasized item which should be clickable 1`] = `
<EuiButton
  fill={true}
  iconSide="right"
  iconType="beaker"
  isDisabled={false}
  onClick={[Function]}
  size="s"
>
  Test
</EuiButton>
`;
